### ISSUE TITLE
Extend "Templates setup window" to show map parts 

### ISSUE THREAD

**Issue by Symbian9** [OpenOrienteering/mapper#1121]
### Actual behavior

Currently *map parts* list exist as separate tool bar. It is not efficient and not best practice.

### Expected behavior

Show *map parts* in "Templates setup window" as separate sub-entries of `- Map -` entry. No need keep separate "Map parts" tool bar (that is unusable).

In future it would be better replace "Templates setup window" by "Project setup window" that will include all templates & map parts. With look like as "Outliner" in Blender or as "Layers" in Inkscape

![pic1](http://i.imgur.com/wIvuwHO.png)

![pic2](http://i.imgur.com/FMjq9Es.png)

**Comment by dg0yt** [comment]
> No need keep separate "Map parts" tool bar (that is unusable).

Why do you think it is not needed or unusable? When using map parts, it seems very important to have a way to quickly *identify* and *switch* the active map part.

**Comment by Symbian9** [comment]
> it seems very important to have a way to quickly
`identify` and `switch` the active map part.

"Map parts" tool bar unusable because it impossible do any other actions using this tool bar such as:

- "Add new part..."
- "Rename current part..."
- "Remove current part"
- "Move selected objects to"
- "Merge this part with"
- "Merge all parts"
- [ ] "Hide current part" (for temporary dissable any pats)

Function "Hide current part" is very important and similar option presented on "Templates setup window" for each *template* or for the whole `- Map -`.

"Move selected objects to" is not needed menu entry at all as we can replace it by doing next steps:

1. "Cut" selected objects on 1st map part;
2. Switch to 2nd map part;
3. "Paste" objects on 2nd map part;

### Project setup window

"Project setup window" should be unifed project organizer dashboard for control all *map parts*, *templates*, etc.

**Comment by dg0yt** [comment]
I see, the tool bar is doing what I expect it to do, but you have very diffent ideas.
This toolbar is intentionally minimal. It is a separate toolbar a) to keep it hidden when there is just a single map part, and b) allow individual placement.

> "Move selected objects to" is not needed menu entry at all as we can replace it by doing next steps:

That is a poor argument. There are lots of features which would not be needed, e.g. Edit > Cut, which is the same as Edit > Copy + Delete.
In addition, the original feature does not change the current map part. You can immediately select the next objects and move them to another part.

**Comment by Symbian9** [comment]
> In addition, the original feature does not change the current map part.

Think, you mean "not switch", because "not change" mostly associated with "not add Undo/Redo steps".

**Comment by lpechacek** [comment]
FWIW, it looks to me like the idea is going in direction how layers are handled in JOSM, which can arbitrarily mix data layers and templates and adjust their opacity and visibility in one docked dialog.

BTW, @Symbian9, what is you backing for statements about said tool efficiency, usability and best practices?

**Comment by Symbian9** [comment]
> what is you backing for statements about said tool efficiency, usability and best practices?

@lpechacek, for short answer: practice and competence. I'm teacher graduated with a degree in technical drawing, informatics & software QA. Cartography is my hobby. Last 7 years working (as volunteer) on testing FOSS CAD, GIS, graphics and DTP software.

**Comment by yevhenmazur** [comment]
Few times per year I'm taking a part in our national workshops about o-mapping, mostly as an OO Mapper expert. Sometimes the students ask me questions that indicates that some things in Mapper are implemented irrational/unclear (for new user). So, in user experience _templates_ and _map parts_ it is just a different kinds of _layers_ in _map project stack_. The user expects that the management of these layers will be carried out from one site and in a similar manner.
The same with issue about [Connect/Unify](https://github.com/OpenOrienteering/mapper/issues/731) tool.

Please note - regional orienteering federation, where the @Symbian9 lives, knows nothing about his o-mapping activity.

**Comment by Symbian9** [comment]
> Please note - regional orienteering federation, where the
@Symbian9 lives, knows nothing about his o-mapping activity.

@yevhenmazur, stop personal attacks! I not work for or affilated with Ukrainian Orienteering Federation. And never claim that.

But, I'm in touch with many ukrainian mappers (not only orienteering mappers), in touch with many orienteering sportsmans and FOSS software developers.

As I said above, I'm graduated with a degree in technical drawing, informatics & software QA. For full activity look on my GitHub profile.

P.S.: @dg0yt, please, remove this bullshit from @yevhenmazur

**Comment by Symbian9** [comment]
> >  No need keep separate "Map parts" tool bar (that is unusable).
> Why do you think it is not needed or unusable? When using map parts, it seems very important to have a way to quickly identify and switch the active map part.

Okay, agree with you. No need remove "Map parts" tool bar.

But need add *map parts* listing on "Templates setup window", where *map parts* should be listed as sub-list of "- Map -" entry.

In OCAD there is "Layout" (layout layers in the map) window, that manage something similar to *map parts* templates that would be printed on resulted map.

- https://www.ocad.com/wiki/ocad/en/index.php?title=Layout
![Layout layers in the map](https://www.ocad.com/wiki/ocad/en/images/8/8e/Order_Layout.png)

In OCAD "Layouts" (layers) are separated from "Background Maps" (in Mapper it called as "Templates")

- https://www.ocad.com/wiki/ocad/en/index.php?title=Background_Map
![Bckground maps](https://www.ocad.com/wiki/ocad/en/images/7/71/ManageBackgroundMap.PNG)

**Comment by michael-77** [comment]
I would agree that it would be helpful to see the map parts in a different way - in particular an ability to show/hide each map part individually would be helpful.

A use case is a large base map with templates in separated layers to allow the use of overlapping areas of the map each with their own copyright, scale and title.

If you could show/hide each map part individually then managing the use of the map would be easier and make sure that changes are only made to the master maps not cut out parts of the map.

See below for sample:

![image](https://user-images.githubusercontent.com/25754945/86500514-8bb0a400-bd89-11ea-9d47-a830eb9cc001.png)

**Comment by jmacura** [comment]
> 
> 
> I would agree that it would be helpful to see the map parts in a different way - in particular an ability to show/hide each map part individually would be helpful.
> 

I have tried to use the map parts only recently and must agree on this. A way how to show/hide individual map parts would be very helpful.

I don't have a strict opinion about whether the map parts should be combined with templates into a "Layers" toolbar, but it's not a bad reasoning.

I would personally like a functionality, which would allow to paint each map part in a single colour. Use case: when combining work from more authors, it would allow me to immediately identify which objects were drawn by the "red" author and which objects were drawn by the "blue" author in the overlapping areas (yes, very conventional colours selected, they are not important, only their mutual contrast is).

**Comment by mlerjen** [comment]
Templates and Map parts are both very useful but also two pairs of shoes I would not like to be cramed into the same overview. The template dialog alone is chronically overloaded...
![image](https://user-images.githubusercontent.com/15611653/86791453-d53f0e80-c069-11ea-8413-fd7e06483aad.png)
... But one could instead think of a map part dialog very similar to the templates dialog. There all map part functions (add, remove, merge, rename) would be integrate, one could activate the active map part but also hide or dim or (monochrome-grayscale) other parts. :-)

**Comment by mlerjen** [comment]
... unless ...

**Comment by mlerjen** [comment]
... unless ...
![image](https://user-images.githubusercontent.com/15611653/86807855-88176880-c07a-11ea-9e22-b16c8c9a914c.png)


... the map entry in the template dialog would be a folder containing all map parts.

**Comment by dg0yt** [comment]
>  a folder containing all map parts.

I think this is what was meant.

For the record:
- Map parts are not layers. There is no up and down, there is no first and last, and that's why there is no easy skipping or tuning during output.
- Map parts were not meant to handle individual print designs. Probably there will be additional requirements related to such use. For example, you may want to have special templates (e.g. logos) switched on and off with selecting a particular print design.

Implementation notes:
- `TemplateTableModel` (just added) needs to be extended to or complemented by a `TemplateTreeModel`, and the dock widget needs to be updated accordingly. Fortunately, this is more feasible now, with the recent introdcution of the proper model interface.
- As as starting point, switching map parts may be implemented as removing/add the elements of these parts from/to the global set of renderables. This is somewhat expensive when switching, but has no overhead in normal editing.

**Comment by Symbian9** [comment]
> Map parts are not layers. There is no up and down, there is no first and last, and that's why there is no easy skipping or tuning during output.

While they are not layers & has no level order, I suggest they still could be represented as a list with automatic ascending (A-Z, 0-9) sorting order basing on map parts names (and refreshing it if some part name changed), as it works in *Blender* for `object`-type objects listing inside "Outliner" window:

![pic.1](https://i.imgur.com/mrJKjGV.png)
